### PR TITLE
expression: support sequence function

### PR DESCRIPTION
**Pull request by AilinKid**
### What problem does this PR solve? 
```
mysql> create sequence seq
Query OK, 0 rows affected (0.00 sec)

mysql> select nextval(seq)
+--------------+
| nextval(seq) |
+--------------+
|            1 |
+--------------+
1 row in set (0.001 sec)

mysql> select lastval(seq)
+--------------+
| lastval(seq) |
+--------------+
|            1 |
+--------------+
1 row in set (0.001 sec)

mysql> select setval(seq, 100)
+------------------+
| setval(seq, 100) |
+------------------+
|              100 |
+------------------+
1 row in set (0.000 sec)
```

### What is changed and how it works?
1: support sequence function `nextval`, `lastval`, `setval`.
2: implement `sequenceState` which could be used to cache sequence lastval.
3: introduce `SequenceSchema` interface to do the sequence schema check in expression package. (If we use information schema, that will cause import cycle problem)

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has introduce `SequenceSchema` interface in util package.

Related changes

 - Need to update the documentation

Release note

 - support sequence function.

### CODE REVIEW

**Comment by AilinKid** [comment]
> No test for those feature?

The sequence allcator is in next PR, the code here is the code in `expression` package.
The function logic test will be in  the PR too. Coming soon!

File: sessionctx/variable/sequence_state.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+)
+
+// SequenceState cache all sequence's latest value accessed by lastval() builtins. It's a session scoped
+// variable, and all public methods of SequenceState are currently-safe.
+type SequenceState struct {
+	mu sync.Mutex
+	// latestValueMap caches the last value obtained by nextval() for each sequence id.
+	latestValueMap map[int64]int64
+	// latestSequence caches the last sequence id that has just been called by nextval().
+	latestSequenceID int64
```
Comment (wshwsh12) [inline sessionctx/variable/sequence_state.go]: Why we need the field?

File: sessionctx/variable/sequence_state.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	"sync"
+)
+
+// SequenceState cache all sequence's latest value accessed by lastval() builtins. It's a session scoped
+// variable, and all public methods of SequenceState are currently-safe.
+type SequenceState struct {
+	mu sync.Mutex
+	// latestValueMap caches the last value obtained by nextval() for each sequence id.
+	latestValueMap map[int64]int64
+	// latestSequence caches the last sequence id that has just been called by nextval().
```
Comment (wshwsh12) [inline sessionctx/variable/sequence_state.go]:
```suggestion
	// latestSequenceID caches the last sequence id that has just been called by nextval().
```

File: sessionctx/variable/sequence_state.go (hunk 1 above)
Comment (AilinKid) [reply sessionctx/variable/sequence_state.go]: Addressed, PTAL

File: expression/builtin_info.go [hunk 3]
```diff
@@ -731,3 +740,144 @@ func (b *builtinTiDBDecodePlanSig) evalString(row chunk.Row) (string, bool, erro
...
+		return 0, false, err
+	}
+	setValue, isNull, err := b.args[1].EvalInt(b.ctx, row)
+	if isNull || err != nil {
+		return 0, isNull, err
+	}
+	return sequence.SetSequenceVal(setValue)
+}
+
+func getSchemaAndSequence(sequenceName string) (schema string, sequence string) {
```
Comment (Deardrops) [inline expression/builtin_info.go]:
```suggestion
func getSchemaAndSequence(sequenceName string) (string, string) {
```

File: sessionctx/variable/sequence_state.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import (
+	"sync"
+)
+
+// SequenceState cache all sequence's latest value accessed by lastval() builtins. It's a session scoped
+// variable, and all public methods of SequenceState are currently-safe.
+type SequenceState struct {
+	mu sync.Mutex
+	// latestValueMap caches the last value obtained by nextval() for each sequence id.
+	latestValueMap map[int64]int64
```
Comment (Deardrops) [inline sessionctx/variable/sequence_state.go]: Consider [sync.Map](https://golang.org/pkg/sync/#Map) and avoid use `sync.Mutex` manually?

Comment (AilinKid) [reply sessionctx/variable/sequence_state.go]: Maybe the `map[int64]int64` is more directly to know what kind of the k&v it is.
Sync.map will store the interface, and transform it, maybe too heavy for int64.

**Comment by Deardrops** [comment]
LGTM

**Pull request merged**